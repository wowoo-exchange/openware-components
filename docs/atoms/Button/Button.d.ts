import '@openware/cryptofont';
import * as React from 'react';
export declare type ButtonClickHandler = (label?: string, e?: React.FormEvent<HTMLInputElement>) => void;
interface ButtonProps {
    /**
     * String that will be displayed as the name on the button
     */
    label: string;
    /**
     * Button click callback
     */
    onClick: ButtonClickHandler;
    /**
     * Additional class name for styling. By default element receives `cr-button` class
     * @default empty
     */
    className?: string;
    /**
     * Remove button margin if true
     */
    noMargin?: boolean;
    /**
     * If true, component will be disabled.
     * @default false
     */
    disabled?: boolean;
    /**
     * Value for defining type of button
     * @default "button"
     */
    type?: string;
}
/**
 * Ð¡r-Button overrides default button.
 */
declare const Button: React.FunctionComponent<ButtonProps>;
export { Button, ButtonProps, };
