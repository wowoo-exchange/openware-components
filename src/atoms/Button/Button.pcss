.cr-button {
  border: none;
  width: 100%;
  color: var(--base-bright);
  background: var(--button-primary);
  padding: var(--gap);
  margin: var(--gap);
  font-weight: bolder;
  text-align: center;
  cursor: pointer;

  &:active {
    background: var(--button-active);
  }

  &:disabled,
  &-disabled {
    background: var(--button-disabled);
    color: var(--button-dark-gray);
    cursor: not-allowed;
  }

  &--no-margin {
    margin: 0;
  }
}
