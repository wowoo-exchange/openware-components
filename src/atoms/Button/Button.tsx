import '@openware/cryptofont';

import classnames from 'classnames';
import * as React from 'react';

export type ButtonClickHandler = (label?: string, e?: React.FormEvent<HTMLInputElement>) => void;

interface ButtonProps {
    /**
     * String that will be displayed as the name on the button
     */
    label: string;
    /**
     * Button click callback
     */
    onClick: ButtonClickHandler;
    /**
     * Additional class name for styling. By default element receives `cr-button` class
     * @default empty
     */
    className?: string;
    /**
     * Remove button margin if true
     */
    noMargin?: boolean;
    /**
     * If true, component will be disabled.
     * @default false
     */
    disabled?: boolean;
    /**
     * Value for defining type of button
     * @default "button"
     */
    type?: string;
}

/**
 * Ð¡r-Button overrides default button.
 */
const Button: React.FunctionComponent<ButtonProps> = props => {
    const { disabled, label, noMargin, className, onClick, type = 'button' } = props;
    const cx = classnames('cr-button', {
        'cr-button--disabled': disabled,
        'cr-button--no-margin': noMargin,
    }, className);
    const value = label.toLocaleUpperCase();
    const handler = !disabled ? (e: React.FormEvent<HTMLInputElement>) => onClick(label, e) : undefined;

    return (
        <input
            readOnly={true}
            type={type}
            className={cx}
            disabled={disabled}
            value={value}
            onClick={handler}
        />
    );
};

export {
    Button,
    ButtonProps,
};
