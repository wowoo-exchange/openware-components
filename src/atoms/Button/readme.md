Example of usage:

```js
const onClick = label => alert(`Clicked: ${label}`);

<div style={{ display: 'flex', justifyContent: 'space-between'}}>
    <Button label="Active" onClick={onClick}/>
    <Button label="Another" onClick={onClick}/>
    <Button label="Disabled" disabled />
</div>
```
