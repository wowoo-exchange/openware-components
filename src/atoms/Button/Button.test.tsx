import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import { SinonSpy, spy } from 'sinon';
import { Button, ButtonProps } from './Button';

const defaults: ButtonProps = {
    label: 'Active',
    onClick: spy(),
};

const setup = (props: Partial<ButtonProps> = {}) =>
    shallow(<Button {...{ ...defaults, ...props }} />);

describe('Button', () => {
    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = setup();
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should have correct className', () => {
        expect(wrapper.hasClass('cr-button')).toBeTruthy();
    });

    it('should pass along supplied className', () => {
        const className = 'some-class';
        wrapper = setup({ className });
        expect(wrapper.hasClass(className)).toBeTruthy();
    });

    it('should pass along supplied type', () => {
        const type = 'submit';
        wrapper = setup({ type });
        expect(wrapper.prop('type')).toEqual('submit');
    });

    it('should set disable className when disabled', () => {
        wrapper = setup({ disabled: true });
        expect(wrapper.prop('disabled')).toBeTruthy();
        expect(wrapper.hasClass('cr-button--disabled')).toBeTruthy();
    });

    it('should set no-margin className', () => {
        wrapper = setup({ noMargin: true });
        expect(wrapper.hasClass('cr-button--no-margin')).toBeTruthy();
    });

    it('should render button with label in UPPER_CASE', () => {
        const expected = defaults.label.toLocaleUpperCase();
        const btn = wrapper.find('input');
        expect(btn.prop('value')).toBe(expected);
    });

    it('should call handler on onClick event with label value', () => {
        wrapper.simulate('click');
        expect((defaults.onClick as SinonSpy).callCount).toBe(1);
        expect(
            (defaults.onClick as SinonSpy).calledOnceWith(defaults.label),
        ).toBeTruthy();
    });
});
