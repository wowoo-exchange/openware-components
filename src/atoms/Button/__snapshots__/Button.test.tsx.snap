// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render 1`] = `
<input
  className="cr-button"
  onClick={[Function]}
  readOnly={true}
  type="button"
  value="ACTIVE"
/>
`;
